hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/stats/SidecarInstrumentationImpl.java

```diff
@@ -0,0 +1,208 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.EnumMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Holds the accumulators needed for the metrics/logging of a sidecar record (corresponding, most
+ * likely, to a record block. There's a separate accumulator for each kind of sidecar record
+ * (bytecode, state change, actions).
+ */
+public record SidecarInstrumentationImpl(
+        @NonNull EnumMap<SidecarType, SidecarAccumulators> accumulators,
+        @NonNull Clock clock,
+        @NonNull AtomicInteger copies)
+        implements SidecarInstrumentation {
+
+    /** Constructs the sidecar instrumentation */
+    public SidecarInstrumentationImpl() {
+        this(new EnumMap<>(SidecarType.class), Clock.systemUTC(), new AtomicInteger());
+        createAccumulators();
+    }
+
+    /** Constructs the sidecar instrumentation, allowing the injection of a `Clock` (for testing) */
+    public SidecarInstrumentationImpl(@NonNull Clock clock) {
+        this(new EnumMap<>(SidecarType.class), clock, new AtomicInteger());
+        createAccumulators();
+    }
+
+    /** Create an accumulator for each kind of sidecar record */
+    private void createAccumulators() {
+        for (final var e : SidecarType.values())
+            accumulators.put(e, new SidecarAccumulators(e.name()));
+    }
+
+    /**
+     * Add a sample of the serialized size of a sidecar transaction, for a particular kind of
+     * sidecar
+     */
+    @Override
+    public void addSample(@NonNull final SidecarType type, long size) {
+        final var acc = accumulators.get(type);
+        acc.serializedSize().addSample(size);
+    }
+
+    /**
+     * Add a sample of the compute duration of a sidecar transaction, for a particular kind of
+     * sidecar
+     */
+    @Override
+    public void addSample(@NonNull final SidecarType type, @NonNull final Duration duration) {
+        final var acc = accumulators.get(type);
+        acc.computeDuration().addSample(duration.toNanos());
+    }
+
+    /**
+     * Add a sample of both serialized size and compute time for a sidecar transaction, for a
+     * particular kind of sidecar
+     */
+    @Override
+    public void addSample(
+            @NonNull final SidecarType type, long size, @NonNull final Duration duration) {
+        final var acc = accumulators.get(type);
+        acc.serializedSize().addSample(size);
+        acc.computeDuration().addSample(duration.toNanos());
+    }
+
+    /**
+     * Add the entire accumulated values of a SidecarInstrumentation to this one (which is a summary
+     * of multiple records
+     */
+    @SuppressWarnings("java:S2445")
+    // sync on method parameter _OK_ because adding samples goes only in _one_ direction ever (from
+    // more specific accumulator to more general accumulator)
+    @Override
+    public void addSamples(@NonNull final SidecarInstrumentation samples) {
+        if (samples instanceof SidecarInstrumentationImpl impl) {
```

Comment:
no else?

Reply:
OK.  Though the only _other_ thing it could be is the noop impl...

Reply:
Doesn't this defy the whole purpose of introducing a noop implementation?

Reply:
no - the noop is really there for the tests. but maybe it is showing that there's something not abstracted properly here.  I'll look.

Reply:
You know, I prefer to leave it this way. The accumulators _are_ an implementation detail that shouldn't be exposed.  And Java doesn't have incomplete types/forward references the way C++ does so the only way to pass a "opaque handle to the internal state" around is to entirely expose it (constrast C++ where you just expose its _name_).  AFAIK.  Correct me?

Reply:
And no else is needed - unless it would be to raise an error.  Which would be kind of pointless.  Reason is: This is dealing with internal details of its own implementation so it doesn't make sense for it to know how to deal with _other_ implementations.  Thanks for bringing this up though.  It caused me to think more carefully about how you can deal with these encapsulation issues in Java (vs, say, C++).  I'm not sure I have a final answer yet ... maybe I'm missing something big ... but I did think about it more.